File: traffic/C-MPI/mplib.c
#include <stdio.h>
#include <mpi.h> 

void mpstart(int *numproc, int *procid)
{ 
  MPI_Init(NULL, NULL);
  MPI_Comm_rank(MPI_COMM_WORLD, procid);
  MPI_Comm_size(MPI_COMM_WORLD, numproc);
}
  
void mpstop(void)
{ 
  MPI_Finalize();
}
 
void mpgsum(int *ival)
{
  int itmp;
 
  MPI_Allreduce(ival, &itmp, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

  *ival = itmp;
} 

void mpscatter(int *inbuff, int *outbuff, int ndata)
{
  MPI_Scatter(inbuff,  ndata, MPI_INT,
	      outbuff, ndata, MPI_INT,
	      0, MPI_COMM_WORLD);
}

void mpsendrecv(int *senddata, int sendproc, int *recvdata, int recvproc)
{
  MPI_Status status;
    
  MPI_Sendrecv(senddata, 1, MPI_INT, sendproc, 1,
	       recvdata, 1, MPI_INT, recvproc, 1,
               MPI_COMM_WORLD, &status);
}
  
double gettime(void)
{ 
  return MPI_Wtime(); 
} 

File: traffic/F-MPI/mplib.f90
module mplib

  use mpi

  implicit none

  integer :: ierror

contains

subroutine mpstart(numproc, procid)

  integer :: numproc, procid

  call MPI_Init(ierror)
  call MPI_Comm_rank(MPI_COMM_WORLD, procid, ierror)
  call MPI_Comm_size(MPI_COMM_WORLD, numproc, ierror)

end subroutine mpstart

subroutine mpstop

  call MPI_Finalize(ierror)

end subroutine mpstop
 
subroutine mpgsum(ival)

  integer :: ival, itmp
 
  call MPI_Allreduce(ival, itmp, 1, MPI_INTEGER, MPI_SUM, &
                     MPI_COMM_WORLD, ierror)

  ival = itmp

end subroutine mpgsum

subroutine mpscatter(inbuff, outbuff, ndata)

  integer :: ndata

  integer, dimension(ndata) :: outbuff
  integer, dimension(:)     :: inbuff

  call MPI_Scatter(inbuff,  ndata, MPI_INTEGER, &
	           outbuff, ndata, MPI_INTEGER, &
                   0, MPI_COMM_WORLD, ierror)

end subroutine mpscatter

subroutine mpsendrecv(senddata, sendproc, recvdata, recvproc)

  integer :: senddata, sendproc, recvdata, recvproc
  integer, dimension(MPI_STATUS_SIZE) :: status

  call MPI_Sendrecv(senddata, 1, MPI_INTEGER, sendproc, 1, &
                    recvdata, 1, MPI_INTEGER, recvproc, 1, &
                    MPI_COMM_WORLD, status, ierror)
 
end subroutine mpsendrecv

double precision function gettime()

  gettime = MPI_Wtime()

end function gettime

end module mplib
